streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/Suppressed.scala

```diff
@@ -98,7 +93,7 @@ object Suppressed {
      * @return size-constrained buffer in terms of the maximum number of bytes it will use.
      * @see [[org.apache.kafka.streams.kstream.Suppressed.BufferConfig.maxBytes]]
      */
-    def maxBytes(byteLimit: Long): BufferConfigJ =
+    def maxBytes(byteLimit: Long): BufferConfigJ[_] =
```

Comment:
@vvcephei this is actually the first thing I tried, but this doesn't compile:


```scala
val bc = Suppressed.BufferConfig
    .maxRecords(1L) // return type of this is `BufferConfig[Nothing]`
    .withMaxBytes(5L) // return type of this is `Nothing`
    .emitEarlyWhenFull() // won't compile
```

IOW, this test will fail: https://github.com/casey-green/kafka/blob/b738a11f4349de9de9686d53ecbe142341c24a4b/streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/kstream/SuppressedTest.scala#L48-L78

You can take it one step further by doing:

```scala
def maxBytes(byteLimit: Long): BufferConfigJ[_ <: BufferConfigJ[_]] = ...
```

But then you can only go three deep with your factory methods:

```scala
val bc = Suppressed.BufferConfig
    .maxRecords(1L) // return type of this is `BufferConfig[BufferConfig[Nothing]]`
    .withMaxBytes(5L) // return type is `BufferConfig[Nothing]`
    .withMaxRecords(5L) // return type is `Nothing`
    .unbounded() // won't compile
```

The other solution I tried is this: https://github.com/casey-green/kafka/compare/casey-green/KAFKA-7778...casey-green:casey-green/KAFKA-7778-scala-211?expand=1

That felt worse to me, but maintains the type parameters.

Reply:
hmm, this is interesting. It compiles and runs/passes for me with Gradle, and I can compile and run `SuppressedTest.scala` in IDEA (it also passes there), but IDEA highlights the lines in question as compile errors.

Is there a chance this is just a problem with the type inference in IDEA? Or am I somehow getting a false negative when I run the tests?

Reply:
gahhhhh IDEA you've led me astray!! 😭 I think you're right... I'll make these changes in the other PR then, thanks @vvcephei

Reply:
FWIW, I have no problem with https://github.com/casey-green/kafka/compare/casey-green/KAFKA-7778...casey-green:casey-green/KAFKA-7778-scala-211?expand=1 .

IIRC, I had basically the same marker interface during development of Suppressed, but removed it just because it wasn't necessary. If it makes the scala adapter possible without a bunch of type-level acrobatics, then it's worth reintroducing it.

The only concern would be source compatibility, but since it's tightening the type, I think it should be compatible with any existing code, even if the buffer config were saved in variables while calling the builder methods with the types declared.

Reply:
Hah! Well, in retrospect, it's probably not good to have source code that _looks_ like it won't compile, either...

I wonder if the `EagerBufferConfig` might be the way to go anyway.

Reply:
Cool, in that case I think I'll go with the `EagerBufferConfig` marker interface approach instead... since I'll be using the scala API and using IDEA 😄 

Updates to the other PR coming up!